File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/utils/GamepadController.java
package org.firstinspires.ftc.teamcode.utils;

import com.qualcomm.robotcore.hardware.Gamepad;

import java.util.HashMap;
import java.util.Map;
import java.util.function.BooleanSupplier;
import java.util.function.Consumer;

public class GamepadController {
    private static final double DEFAULT_THRESHOLD = 0.1;
    private final Gamepad gamepad;
    private final Map<BooleanSupplier, Boolean> prevButtonStates = new HashMap<>();
    private final Map<BooleanSupplier, Runnable> pressCallbacks = new HashMap<>();
    private final Map<BooleanSupplier, Runnable> releaseCallbacks = new HashMap<>();
    private final Map<BooleanSupplier, Consumer<Double>> whilePressedCallbacks = new HashMap<>();
    private double threshold = DEFAULT_THRESHOLD;
    private long lastOperationTime = System.currentTimeMillis();

    public GamepadController(Gamepad gamepad) {
        this.gamepad = gamepad;
    }

    public GamepadController setThreshold(double threshold) {
        this.threshold = Math.max(0.0, Math.min(1.0, threshold));
        return this;
    }

    public long getNoOperationTime() {
        return System.currentTimeMillis() - lastOperationTime;
    }

    public GamepadController onPressed(BooleanSupplier buttonSupplier, Runnable callback) {
        pressCallbacks.put(buttonSupplier, callback);
        prevButtonStates.putIfAbsent(buttonSupplier, false);
        return this;
    }

    public GamepadController onPressed(ButtonType button, Runnable callback) {
        return onPressed(button(button), callback);
    }

    public GamepadController onReleased(BooleanSupplier buttonSupplier, Runnable callback) {
        releaseCallbacks.put(buttonSupplier, callback);
        prevButtonStates.putIfAbsent(buttonSupplier, false);
        return this;
    }

    public GamepadController onReleased(ButtonType button, Runnable callback) {
        return onReleased(button(button), callback);
    }

    public GamepadController whilePressed(BooleanSupplier buttonSupplier, Consumer<Double> callback) {
        whilePressedCallbacks.put(buttonSupplier, callback);
        prevButtonStates.putIfAbsent(buttonSupplier, false);
        return this;
    }

    public GamepadController whilePressed(ButtonType button, Consumer<Double> callback) {
        return whilePressed(button(button), callback);
    }

    public void update() {
        for (Map.Entry<BooleanSupplier, Boolean> entry : prevButtonStates.entrySet()) {
            BooleanSupplier buttonSupplier = entry.getKey();
            boolean prevState = entry.getValue();
            boolean currentState = buttonSupplier.getAsBoolean();

            if (!prevState && currentState) {
                lastOperationTime = System.currentTimeMillis();
                Runnable callback = pressCallbacks.get(buttonSupplier);
                if (callback != null) {
                    callback.run();
                }
            }

            if (prevState && !currentState) {
                Runnable callback = releaseCallbacks.get(buttonSupplier);
                if (callback != null) {
                    callback.run();
                }
            }

            if (currentState) {
                Consumer<Double> callback = whilePressedCallbacks.get(buttonSupplier);
                if (callback != null) {
                    callback.accept(1.0);
                }
            }

            entry.setValue(currentState);
        }
    }

    public BooleanSupplier button(ButtonType button) {
        switch (button) {
            case A:
                return () -> gamepad.a;
            case B:
                return () -> gamepad.b;
            case X:
                return () -> gamepad.x;
            case Y:
                return () -> gamepad.y;
            case DPAD_UP:
                return () -> gamepad.dpad_up;
            case DPAD_DOWN:
                return () -> gamepad.dpad_down;
            case DPAD_LEFT:
                return () -> gamepad.dpad_left;
            case DPAD_RIGHT:
                return () -> gamepad.dpad_right;
            case LEFT_BUMPER:
                return () -> gamepad.left_bumper;
            case RIGHT_BUMPER:
                return () -> gamepad.right_bumper;
            case LEFT_STICK_BUTTON:
                return () -> gamepad.left_stick_button;
            case RIGHT_STICK_BUTTON:
                return () -> gamepad.right_stick_button;
            case BACK:
                return () -> gamepad.back;
            case START:
                return () -> gamepad.start;
            case GUIDE:
                return () -> gamepad.guide;
            default:
                return () -> false;
        }
    }

    public BooleanSupplier trigger(TriggerType trigger) {
        switch (trigger) {
            case LEFT_TRIGGER:
                return () -> gamepad.left_trigger > threshold;
            case RIGHT_TRIGGER:
                return () -> gamepad.right_trigger > threshold;
            default:
                return () -> false;
        }
    }

    public double getTriggerValue(TriggerType trigger) {
        switch (trigger) {
            case LEFT_TRIGGER:
                return gamepad.left_trigger;
            case RIGHT_TRIGGER:
                return gamepad.right_trigger;
            default:
                return 0.0;
        }
    }

    public double getJoystickX(JoystickType joystick) {
        switch (joystick) {
            case LEFT:
                return gamepad.left_stick_x;
            case RIGHT:
                return gamepad.right_stick_x;
            default:
                return 0.0;
        }
    }

    public double getJoystickY(JoystickType joystick) {
        switch (joystick) {
            case LEFT:
                return -gamepad.left_stick_y; // Inverted to match standard coordinate system
            case RIGHT:
                return -gamepad.right_stick_y; // Inverted to match standard coordinate system
            default:
                return 0.0;
        }
    }

    public BooleanSupplier joystickMoved(JoystickType joystick) {
        return () -> {
            double x = 0.0;
            double y = 0.0;

            switch (joystick) {
                case LEFT:
                    x = gamepad.left_stick_x;
                    y = gamepad.left_stick_y;
                    break;
                case RIGHT:
                    x = gamepad.right_stick_x;
                    y = gamepad.right_stick_y;
                    break;
            }
            return Math.sqrt(x * x + y * y) > threshold;
        };
    }

    public boolean isRumbling() {
        return gamepad.isRumbling();
    }

    public void rumble(int durationMs) {
        gamepad.rumble(durationMs);
    }

    public void rumble(double leftPower, double rightPower, int durationMs) {
        gamepad.rumble(leftPower, rightPower, durationMs);
    }

    public enum ButtonType {
        A, B, X, Y,
        DPAD_UP, DPAD_DOWN, DPAD_LEFT, DPAD_RIGHT,
        LEFT_BUMPER, RIGHT_BUMPER,
        LEFT_STICK_BUTTON, RIGHT_STICK_BUTTON,
        BACK, START, GUIDE
    }

    public enum TriggerType {
        LEFT_TRIGGER, RIGHT_TRIGGER
    }

    public enum JoystickType {
        LEFT, RIGHT
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/utils/GamepadController.kt
package org.firstinspires.ftc.teamcode.utils

import com.qualcomm.robotcore.hardware.Gamepad
import java.util.function.BooleanSupplier
import java.util.function.Consumer
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow
import kotlin.math.sqrt

/**
 * An enhanced controller for FTC Gamepad that provides event-based handling
 * for button presses and releases, along with other utility functions.
 */
class GamepadController
/**
 * Creates a new GamepadController for the given gamepad.
 *
 * @param gamepad The gamepad to monitor
 */(// The gamepad to monitor
    private val gamepad: Gamepad
) {
    // Maps to store the previous state of buttons
    private val prevButtonStates: MutableMap<BooleanSupplier, Boolean> = HashMap()

    // Maps to store callbacks for button events
    private val pressCallbacks: MutableMap<BooleanSupplier, Runnable> = HashMap()
    private val releaseCallbacks: MutableMap<BooleanSupplier, Runnable> = HashMap()
    private val whilePressedCallbacks: MutableMap<BooleanSupplier, Consumer<Double>> = HashMap()

    private var threshold = DEFAULT_THRESHOLD

    var lastOperationTime: Long = System.currentTimeMillis()
    /**
     * Sets the threshold for considering analog inputs as "pressed".
     *
     * @param threshold The threshold value (0.0 to 1.0)
     * @return This controller for method chaining
     */
    fun setThreshold(threshold: Double): GamepadController {
        this.threshold = max(0.0, min(1.0, threshold))
        return this
    }

    fun getNoOperationTime(): Long {
        return System.currentTimeMillis() - lastOperationTime
    }

    /**
     * Registers a callback to be called when a button is pressed.
     *
     * @param buttonSupplier The function that returns the button state
     * @param callback The callback to execute when the button is pressed
     * @return This controller for method chaining
     */
    fun onPressed(buttonSupplier: BooleanSupplier, callback: Runnable): GamepadController {
        pressCallbacks[buttonSupplier] = callback
        prevButtonStates.putIfAbsent(buttonSupplier, false)
        return this
    }

    /**
     * Registers a callback to be called when a button is pressed.
     * Simplified version that takes a ButtonType directly.
     *
     * @param button The button type to monitor
     * @param callback The callback to execute when the button is pressed
     * @return This controller for method chaining
     */
    fun onPressed(button: ButtonType, callback: Runnable): GamepadController {
        return onPressed(button(button), callback)
    }

    /**
     * Registers a callback to be called when a button is released.
     *
     * @param buttonSupplier The function that returns the button state
     * @param callback The callback to execute when the button is released
     * @return This controller for method chaining
     */
    fun onReleased(buttonSupplier: BooleanSupplier, callback: Runnable): GamepadController {
        releaseCallbacks[buttonSupplier] = callback
        prevButtonStates.putIfAbsent(buttonSupplier, false)
        return this
    }

    /**
     * Registers a callback to be called when a button is released.
     * Simplified version that takes a ButtonType directly.
     *
     * @param button The button type to monitor
     * @param callback The callback to execute when the button is released
     * @return This controller for method chaining
     */
    fun onReleased(button: ButtonType, callback: Runnable): GamepadController {
        return onReleased(button(button), callback)
    }

    /**
     * Registers a callback to be called continuously while a button is pressed.
     *
     * @param buttonSupplier The function that returns the button state
     * @param callback The callback to execute while the button is pressed,
     * receives a value from 0.0 to 1.0 for analog inputs
     * @return This controller for method chaining
     */
    fun whilePressed(
        buttonSupplier: BooleanSupplier,
        callback: Consumer<Double>
    ): GamepadController {
        whilePressedCallbacks[buttonSupplier] = callback
        prevButtonStates.putIfAbsent(buttonSupplier, false)
        return this
    }

    /**
     * Registers a callback to be called continuously while a button is pressed.
     * Simplified version that takes a ButtonType directly.
     *
     * @param button The button type to monitor
     * @param callback The callback to execute while the button is pressed
     * @return This controller for method chaining
     */
    fun whilePressed(button: ButtonType, callback: Consumer<Double>): GamepadController {
        return whilePressed(button(button), callback)
    }

    /**
     * Updates the controller state and executes any registered callbacks.
     * Should be called once per loop cycle.
     */
    fun update() {
        // Check button press/release events
        for (entry in prevButtonStates.entries) {
            val buttonSupplier = entry.key
            val prevState = entry.value
            val currentState = buttonSupplier.asBoolean

            // Check for button press, only when button pressed update lastOperationTime
            if (!prevState && currentState) {
                lastOperationTime = System.currentTimeMillis()
                val callback = pressCallbacks[buttonSupplier]
                callback?.run()
            }

            // Check for button release
            if (prevState && !currentState) {
                val callback = releaseCallbacks[buttonSupplier]
                callback?.run()
            }

            // Check for while pressed
            if (currentState) {
                val callback = whilePressedCallbacks[buttonSupplier]
                callback?.accept(1.0)
            }

            // Update previous state
            entry.setValue(currentState)
        }
    }

    /**
     * Creates a supplier for a gamepad button.
     */
    fun button(button: ButtonType): BooleanSupplier {
        return when (button) {
            ButtonType.A -> BooleanSupplier { gamepad.a }
            ButtonType.B -> BooleanSupplier { gamepad.b }
            ButtonType.X -> BooleanSupplier { gamepad.x }
            ButtonType.Y -> BooleanSupplier { gamepad.y }
            ButtonType.DPAD_UP -> BooleanSupplier { gamepad.dpad_up }
            ButtonType.DPAD_DOWN -> BooleanSupplier { gamepad.dpad_down }
            ButtonType.DPAD_LEFT -> BooleanSupplier { gamepad.dpad_left }
            ButtonType.DPAD_RIGHT -> BooleanSupplier { gamepad.dpad_right }
            ButtonType.LEFT_BUMPER -> BooleanSupplier { gamepad.left_bumper }
            ButtonType.RIGHT_BUMPER -> BooleanSupplier { gamepad.right_bumper }
            ButtonType.LEFT_STICK_BUTTON -> BooleanSupplier { gamepad.left_stick_button }
            ButtonType.RIGHT_STICK_BUTTON -> BooleanSupplier { gamepad.right_stick_button }
            ButtonType.BACK -> BooleanSupplier { gamepad.back }
            ButtonType.START -> BooleanSupplier { gamepad.start }
            ButtonType.GUIDE -> BooleanSupplier { gamepad.guide }
            else -> BooleanSupplier { false }
        }
    }

    /**
     * Creates a supplier for a gamepad trigger.
     */
    fun trigger(trigger: TriggerType): BooleanSupplier {
        return when (trigger) {
            TriggerType.LEFT_TRIGGER -> BooleanSupplier { gamepad.left_trigger > threshold }
            TriggerType.RIGHT_TRIGGER -> BooleanSupplier { gamepad.right_trigger > threshold }
            else -> BooleanSupplier { false }
        }
    }

    /**
     * Gets the raw value of a trigger (0.0 to 1.0).
     */
    fun getTriggerValue(trigger: TriggerType): Double {
        return when (trigger) {
            TriggerType.LEFT_TRIGGER -> gamepad.left_trigger.toDouble()
            TriggerType.RIGHT_TRIGGER -> gamepad.right_trigger.toDouble()
            else -> 0.0
        }
    }

    /**
     * Gets the X value of a joystick (-1.0 to 1.0).
     */
    fun getJoystickX(joystick: JoystickType): Double {
        return when (joystick) {
            JoystickType.LEFT -> gamepad.left_stick_x.toDouble()
            JoystickType.RIGHT -> gamepad.right_stick_x.toDouble()
            else -> 0.0
        }
    }

    /**
     * Gets the Y value of a joystick (-1.0 to 1.0).
     */
    fun getJoystickY(joystick: JoystickType): Double {
        return when (joystick) {
            JoystickType.LEFT -> -gamepad.left_stick_y.toDouble() // Inverted to match standard coordinate system
            JoystickType.RIGHT -> -gamepad.right_stick_y.toDouble() // Inverted to match standard coordinate system
            else -> 0.0
        }
    }

    /**
     * Checks if a joystick is pushed beyond the threshold in any direction.
     */
    fun joystickMoved(joystick: JoystickType): BooleanSupplier {
        return BooleanSupplier {
            var x = 0.0
            var y = 0.0

            when (joystick) {
                JoystickType.LEFT -> {
                    x = gamepad.left_stick_x.toDouble()
                    y = gamepad.left_stick_y.toDouble()
                }

                JoystickType.RIGHT -> {
                    x = gamepad.right_stick_x.toDouble()
                    y = gamepad.right_stick_y.toDouble()
                }
            }
            sqrt(x * x + y * y) > threshold
        }
    }

    val isRumbling: Boolean
        /**
         * Returns whether the gamepad is currently rumbling.
         */
        get() = gamepad.isRumbling

    /**
     * Sets the gamepad to rumble for a specified duration.
     * @param durationMs The duration in milliseconds
     */
    fun rumble(durationMs: Int) {
        gamepad.rumble(durationMs)
    }

    /**
     * Sets the gamepad to rumble.
     *
     * @param leftPower Left rumble motor power (0.0 to 1.0)
     * @param rightPower Right rumble motor power (0.0 to 1.0)
     * @param durationMs The duration in milliseconds
     */
    fun rumble(leftPower: Double, rightPower: Double, durationMs: Int) {
        gamepad.rumble(leftPower, rightPower, durationMs)
    }

    /**
     * Enum for button types.
     */
    enum class ButtonType {
        A, B, X, Y,
        DPAD_UP, DPAD_DOWN, DPAD_LEFT, DPAD_RIGHT,
        LEFT_BUMPER, RIGHT_BUMPER,
        LEFT_STICK_BUTTON, RIGHT_STICK_BUTTON,
        BACK, START, GUIDE
    }

    /**
     * Enum for trigger types.
     */
    enum class TriggerType {
        LEFT_TRIGGER, RIGHT_TRIGGER
    }

    /**
     * Enum for joystick types.
     */
    enum class JoystickType {
        LEFT, RIGHT
    }

    companion object {
        // Threshold for considering analog inputs as "pressed"
        private const val DEFAULT_THRESHOLD = 0.1
    }
}